server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -188,11 +215,12 @@ public DataStream(
         this.indexMode = indexMode;
         this.lifecycle = lifecycle;
         this.failureStoreEnabled = failureStoreEnabled;
+        assert backingIndices.indices.isEmpty() == false;
+        assert assertConsistent(backingIndices.indices);
```

Comment:
Maybe orthogonal to this PR, but should we assert that the failure stores provided are consistent as well?

Considering that method builds a set of the index names, but we also make a set of the failure store names lazily, should we just eagerly create that set instead of doing it again when needed?

Reply:
I hope you do not mind me reopening this topic for a quick remark.

The asserts are not executed on release code, so the sets are not created at all.
With this change, a released version will create a lot more sets than before, not only for every data stream but also for every rebuild of the data stream object, which I think is a lot. Also we do not even use it that much, so for most cases it will go to waste. 

Considering the above, I think that lazily creating the set for both backing indices and failure stores remains the way to go. What do you think?

Reply:
Ah I think I didn't take into account that the consistency assertions were actual assertions. Good catch :). I agree with you that lazily creating them makes more sense. I'll wait to make changes until @jbaiera has had a chance to respond to this as well.

Reply:
I agree with @gmarouli on this one, managed to ignore the assert keyword in my thinking